libpod/healthcheck.go

```diff
@@ -146,9 +146,13 @@ func (c *Container) runHealthCheck(ctx context.Context, isStartup bool) (define.
 		hcErr = fmt.Errorf("healthcheck command exceeded timeout of %s", c.HealthCheckConfig().Timeout.String())
 	}
 
+	if exitCode != 0 && c.ensureCurrentState(define.ContainerStateStopped, define.ContainerStateStopping, define.ContainerStateExited) {
+		hcResult = define.HealthCheckContainerStopped
+	}
+
 	hcl := newHealthCheckLog(timeStart, timeEnd, returnCode, eventLog)
 
-	healthCheckResult, err := c.updateHealthCheckLog(hcl, inStartPeriod, isStartup)
+	healthCheckResult, err := c.updateHealthCheckLog(hcl, hcResult, inStartPeriod, isStartup)
```

Comment:
this ensureCurrentState takes the lock and updateHealthCheckLog takes the lock again, this is inefficient and should not happen.
I would say remove both locks from the functions and take on lock here in the caller instead.